tests/failpoints_cases/test_snap.rs

```diff
@@ -66,3 +74,111 @@ fn test_overlap_cleanup() {
     }
     fail::remove(gen_snapshot_fp);
 }
+
+pub struct SnapshotNotifier {
+    notifier: Mutex<Sender<()>>,
+    pending_notify: AtomicUsize,
+    ready_notify: Arc<AtomicBool>,
+}
+
+impl SnapshotNotifier {
+    pub fn new(notifier: Sender<()>, ready_notify: Arc<AtomicBool>) -> SnapshotNotifier {
+        SnapshotNotifier {
+            notifier: Mutex::new(notifier),
+            ready_notify: ready_notify,
+            pending_notify: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl Filter<RaftMessage> for SnapshotNotifier {
+    fn before(&self, msgs: &mut Vec<RaftMessage>) -> Result<()> {
+        for msg in msgs.iter() {
+            if msg.get_message().get_msg_type() == MessageType::MsgSnapshot
+                && self.ready_notify.load(Ordering::SeqCst)
+            {
+                self.pending_notify.fetch_add(1, Ordering::SeqCst);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn after(&self, _: Result<()>) -> Result<()> {
```

Comment:
Should we check message type here?

Reply:
No, we have checked in the `before` and we want to send notifications after sending messages.